js/src/api/transport/ws/ws.js

```diff
@@ -194,31 +202,66 @@ export default class Ws extends JsonRpcBase {
     }, 50);
   }
 
+  _extract = (result) => {
+    // initial pubsub ACK
+    if (result.id && this._messages[result.id].subscription) {
+      this._messages[result.id].method === 'eth_subscribe'
+        ? this._subscriptions['eth_subscription'][result.result] = result.id
+        : this._subscriptions['parity_subscription'][result.result] = result.id;
+      // resolve promise with messageId because subId's can collide (eth/parity)
+      this._messages[result.id].resolve(result.id);
+      // save subId for unsubscribing later
+      this._messages[result.id]['subId'] = result.result;
+      return this._messages[result.id];
+    // normal message
+    } else if (result.id) {
+      return this._messages[result.id];
+    // pubsub format
+    } else if (result.method.includes('subscription')) {
+      if (this._messages[this._subscriptions[result.method][result.params.subscription]]) {
+        return this._messages[this._subscriptions[result.method][result.params.subscription]];
+      } else {
+        throw Error(`Received Subscription which is already unsubscribed ${JSON.stringify(result)}`);
+      }
+    } else {
+      throw Error(`Unknown message format: No ID or subscription ${JSON.stringify(result)}`);
+    }
+  }
+
   _onMessage = (event) => {
     try {
       const result = JSON.parse(event.data);
-      const { method, params, json, resolve, reject } = this._messages[result.id];
+      const { method, params, json, resolve, reject, callback } = this._extract(result);
 
       Logging.send(method, params, { json, result });
 
-      if (result.error) {
+      if (result.error || (result.params && result.params.error)) {
         this.error(event.data);
 
+        if (result.params.error) {
+          result[error] = result.params.error;
```

Comment:
I'm pretty sure that's not what you want (should be `result.error =` instead).

Actually I think it might be better to write it like this:
```js
result.error = (result.params && result.params.error) || result.error;
if (result.error) {
  ...
}
```